hamlet/theses/document.py

```diff
@@ -0,0 +1,67 @@
+from chardet.universaldetector import UniversalDetector
+from django.core.exceptions import ValidationError
+import docx
+import magic
+
+
+def factory(fp):
+    """Factory for creating a document object.
+
+    A Document must implement a `words` property that consists of a list
+    of words in that document.
+
+    :param fp: `django.core.files.uploadedfile.UploadedFile`
+    :return: document object
+    """
+    mimetype = magic.from_buffer(fp.read(1024), mime=True)
+    if mimetype == "text/plain":
+        return TextDocument(fp)
+    elif mimetype == "application/zip":
+        # docx is a zip archive, so this is what libmagic will see it as
+        try:
+            return DocxDocument(fp)
+        except:
+            raise ValidationError("Invalid document type")
+    else:
+        raise ValidationError("Invalid document type")
+
+
+class TextDocument:
+    """Document object for representing a text document."""
+    def __init__(self, doc):
+        self.doc = doc
+        self._words = []
+        self._encoding = ""
+
+    @property
+    def words(self):
+        if not self._words:
+            for line in self.doc:
+                self._words.extend(line.decode(self.encoding).strip().split())
+        return self._words
+
+    @property
+    def encoding(self):
+        if not self._encoding:
+            detector = UniversalDetector()
+            for line in self.doc:
+                detector.feed(line)
+                if detector.done:
+                    break
+            detector.close()
+            self._encoding = detector.result['encoding']
+        return self._encoding
+
+
+class DocxDocument:
+    """Document object for representing a DOCX document."""
+    def __init__(self, doc):
+        self._words = []
+        self.doc = docx.Document(doc)
```

Comment:
I take it .docx handles all encoding issues internally?

Reply:
Yeah, by the time we get text out of it, it's already unicode. I will say that if we decide we don't like how the library I'm using (it's the only one I could find) is performing it probably wouldn't be totally onerous to roll our own. Our needs for pulling text out of the document are very minimal.